### PR TITLE
[WIP] Add order field to reference objects

### PR DESCRIPTION
**Pull request by tompollard**
# Background

Project descriptions include a "References" section that lists relevant papers, websites, etc. An example of a references section for a demo project is shown below (screenshot from http://localhost:8000/content/demoecg/10.5.24/):

Screen Shot 2022-06-02 at 14 04 47

# Problem

As discussed in https://github.com/MIT-LCP/physionet-build/issues/1557 and https://github.com/MIT-LCP/physionet-build/issues/1105, the way we manage these references is extremely buggy.  The problem is that there is no "order" field associated with the references. Instead the references are ordered by their primary key!

For both `ActiveProject` and `ArchivedProject`, the references are stored using the `Reference` class:

https://github.com/MIT-LCP/physionet-build/blob/1dc09c6ce67ad6d77c76a1ebe16a2dcdee65c883/physionet-django/project/modelcomponents/metadata.py#L575-L587

For `PublishedProject`, the references are copied to the `PublishedReference` class:

https://github.com/MIT-LCP/physionet-build/blob/1dc09c6ce67ad6d77c76a1ebe16a2dcdee65c883/physionet-django/project/modelcomponents/metadata.py#L601-L610

# Fixing the problem

This pull request:
- adds an `order` attribute to the `Reference` and `PublishedReference` classes.
- adds a migration that (a) adds the order field (2) migrates the order of existing references to the order field
- updates the view/templates to use the order field to display references.

What's needed before merging:
- We need to update the forms displayed in the following template so that newly created projects record `order` (instead of the ID): https://github.com/MIT-LCP/physionet-build/blob/dev/physionet-django/project/templates/project/project_content.html

The current formset is:

https://github.com/MIT-LCP/physionet-build/blob/1dc09c6ce67ad6d77c76a1ebe16a2dcdee65c883/physionet-django/project/forms.py#L658-L691

The formset is rendered by a generic "item_list" template:
https://github.com/MIT-LCP/physionet-build/blob/dev/physionet-django/project/templates/project/item_list.html

### CODE REVIEW

**Comment by tompollard** [comment]
@alistairewj is someone on your side able to help with the following point?:

> We need to update the form displayed in the following template so that newly created projects record "`order` (instead of the ID): https://github.com/MIT-LCP/physionet-build/blob/dev/physionet-django/project/templates/project/project_content.html

File: physionet-django/project/modelcomponents/metadata.py [hunk 1]
```diff
@@ -581,12 +581,13 @@ class Reference(models.Model):
...
     object_id = models.PositiveIntegerField()
     project = GenericForeignKey('content_type', 'object_id')
+    order = models.PositiveIntegerField(null=True)
 
     description = models.CharField(max_length=1000)
 
     class Meta:
         default_permissions = ()
-        unique_together = (('description', 'content_type', 'object_id'),)
+        unique_together = (('description', 'content_type', 'order'),)
```
Comment (superryeti) [inline physionet-django/project/modelcomponents/metadata.py:590]: @tompollard  I am not sure if the unique is currently setup in the best possible way. Our current unique setup will cause false duplicate error in case where the reference and order of reference is same in two different projects(of same project type).

For example

Project A ( contentType=database)

| Order | reference | uniqueKey      |
|-------|-----------|----------------|
| 1     | 'a'       | (27, 'a', 1)   |
| 2     | 'b'       | (27, 'b', 2)   |
| 3     | 'c'       | (27, 'c', 3)   |
| 4     | 'd'       | (27, 'd', 4)   |


Project B ( contentType=database)

| Order | reference | uniqueKey      |
|-------|-----------|----------------|
| 1     | 'aa'       | (27, 'aa', 1)  |
| 2     | 'b'       | (27, 'b', 2)   |
| 3     | 'cc'      | (27, 'cc', 3)  |
| 4     | 'dd'      | (27, 'dd', 4)  |


as you can see we will have a conflict on the 2nd reference.


I think project might be better instead of content_type.

Comment (superryeti) [reply physionet-django/project/modelcomponents/metadata.py:590]: maybe not `project` exactly, as the generic foreign key throws error when trying to use in `unique_together`

Comment (superryeti) [reply physionet-django/project/modelcomponents/metadata.py:590]: not sure if adding object_id to unique_together is a good idea as that would most likely allow is to insert duplicates

for example
| Order | reference | uniqueKey                | content_id |
|-------|-----------|--------------------------|------------|
| 1     | 'a'       | (27, 'a', 1,11)    | 11   |
| 2     | 'b'       | (27, 'b', 2,12)    | 12   |
| 3     | 'c'       | (27, 'c', 3,13)    | 13   |
| 4     | 'd'       | (27, 'd', 4,14)    | 14   |
| 5     | 'a'       | (27, 'a', 5,15)    | 15   |


Although the front end will not send the duplicate reference, but updating the unique_together this way would make zero sense, as it would allow duplicate in database

Comment (tompollard) [reply physionet-django/project/modelcomponents/metadata.py:590]: > Our current unique setup will cause false duplicate error in case where the reference and order of reference is same in two different projects(of same project type).

Yes, good catch, you're right! This doesn't make sense. 

> I think project might be better instead of content_type ... maybe not project exactly, as the generic foreign key throws error when trying to use in unique_together

I agree, `content_type` doesn't add anything useful here. I also agree that project would be a good alternative (and it's a shame that generic foreign keys make this difficult).

> not sure if adding object_id to unique_together is a good idea as that would most likely allow is to insert duplicates

Unless I'm missing something, I think it is fine to switch `object_id` with `content_type`. It is true that we may end up with duplicate reference strings in the database (slightly inefficient in terms of storage) but I don't think this is too much to worry about.

In summary, I think you should be good to change `content_type` to `object_id`.

Comment (superryeti) [reply physionet-django/project/modelcomponents/metadata.py:590]: Alright, i think that makes sense `object_id` instead of `content_type` should work. The object ID is a integer that increments automatically. So it should help us prevent duplicate across different project.(and the order should help fix the reference error that we were trying to fix)

I will do more tests, and submit a PR in the monring

Thanks @tompollard
